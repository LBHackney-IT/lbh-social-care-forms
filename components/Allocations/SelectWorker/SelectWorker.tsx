import cx from 'classnames';
import { Worker } from 'types';

const ResultEntry = ({
  worker,
  callback,
}: {
  worker: Worker;
  callback: (workerId: string) => void;
}): React.ReactElement => {
  const { id, firstName, lastName, allocationCount } = worker;

  return (
    <>
      <tr className={cx('govuk-table__row')}>
        <td className="govuk-table__cell">
          <div className="govuk-radios__item">
            <input
              id={String(id)}
              name="workerId"
              type="radio"
              value={id}
              onChange={(e) => {
                callback(e.target.value);
              }}
              className={`govuk-radios__input`}
            />
            <label
              className="govuk-label govuk-radios__label"
              htmlFor={String(id)}
            >
              <span className="govuk-visually-hidden">
                {`${firstName} ${lastName}`}
              </span>
            </label>
          </div>
        </td>
        <td data-testid="workerName" className="govuk-table__cell">
          {`${firstName} ${lastName}`}
        </td>
        <td className="govuk-table__cell">
          <label data-testid="allocationCount" htmlFor={String(id)}>
            {allocationCount}
          </label>
        </td>
      </tr>
    </>
  );
};

const SelectWorker = ({
  records,
  callback,
}: {
  records: Worker[];
  callback: any;
}): React.ReactElement => {
  return (
    <table className="govuk-table lbh-table" data-testid="residents-table">
      <thead className="govuk-table__head">
        <tr className="govuk-table__row">
          <th scope="col" className="govuk-table__header">
            Select
          </th>
          <th scope="col" className="govuk-table__header">
            Name
          </th>
          <th scope="col" className="govuk-table__header">
            Total allocations
          </th>
          <th scope="col" className="govuk-table__header" />
        </tr>
      </thead>

      <tbody className="govuk-table__body">
        {records.map((result) => (
          <ResultEntry key={result.id} worker={result} callback={callback} />
        ))}
      </tbody>
    </table>
  );
};

export default SelectWorker;
