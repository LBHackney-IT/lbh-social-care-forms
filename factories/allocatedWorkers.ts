import { Factory } from 'fishery';

import { Allocation } from 'types';

export const allocationFactory = Factory.define<Allocation>(({ sequence }) => ({
  id: sequence,
  allocatedWorkerTeam: 'Safeguarding and Reviewing Team',
  allocatedWorker: 'Officer Name',
  allocationEndDate: '2019-03-28 00:00:00',
  allocationStartDate: '2019-03-28 00:00:00',
  workerType: 'Consultant Social Worker',
  personId: 1234,
  caseStatus: 'Open',
  personName: 'foo',
  personAddress: 'the address',
  personDateOfBirth: '2020-03-20',
  ragRating: 'amber',
}));

export const mockedAllocation = allocationFactory.build();
export const mockedAllocations = [allocationFactory.build()];
