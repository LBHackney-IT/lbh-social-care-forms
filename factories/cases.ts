import { Factory } from 'fishery';

import { Case } from 'types';

export const caseFactory = Factory.define<Case>(({ sequence }) => ({
  recordId: sequence.toString(),
  personId: 123,
  dateOfEvent: '25/10/2020 13:49:43',
  formName: 'foorm',
  officerEmail: 'Fname.Lname@hackney.gov.uk',
  caseFormData: {
    note: 'I am the note',
    first_name: 'Foo',
    last_name: 'Bar',
    timestamp: '11/03/2021 16:51:56',
    mosaic_id: 123,
    form_name_overall: 'foo',
    worker_email: 'worker@hackney.gov.uk',
    form_name: 'Foo bar',
    context_flag: 'A',
    date_of_event: '25/10/2020 13:49:43',
  },
  deleted: false,
  isImported: false,
}));

export const mockedNote = caseFactory.build({ caseFormUrl: 'https://foo.bar' });
export const mockedAllocationNote = caseFactory.build({
  formName: 'Worker allocated',
  caseFormData: {
    allocation_id: 321,
    created_by: 'foo@hackney.gov.uk',
    form_name_overall: 'API_Allocation',
    form_name: 'Worker allocated',
  },
});
export const mockedDeallocationNote = caseFactory.build({
  formName: 'Worker allocated',
  caseFormData: {
    allocation_id: 321,
    created_by: 'foo@hackney.gov.uk',
    deallocation_reason: 'deallocated because',
    form_name_overall: 'API_Deallocation',
    form_name: 'Worker deallocated',
  },
});
export const mockedCaseNote = caseFactory.build({
  caseFormData: {
    form_name_overall: 'ASC_case_note',
    case_note_title: 'i am a case title',
  },
});
export const mockedDeletedCaseNote = caseFactory.build({
  caseFormData: {
    form_name_overall: 'ASC_case_note',
    case_note_title: 'i am a case title',
  },
  formName: 'adult-case-note',
  deleted: true,
});
export const mockedWarningNoteCase = caseFactory.build({
  formName: 'Warning Note',
  caseFormData: {
    form_name_overall: 'API_WarningNote',
    warning_note_id: 456,
  },
});
