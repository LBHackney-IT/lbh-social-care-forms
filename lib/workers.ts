import axios from 'axios';

import { Worker } from 'types';

const ENDPOINT_API = process.env.ENDPOINT_API;
const AWS_KEY = process.env.AWS_KEY;
import { getQueryString } from 'utils/urls';

const headers = {
  'x-api-key': AWS_KEY,
};

export const getWorkers = async (
  params?: Record<string, unknown>
): Promise<Worker[]> => {
  const urlParameters = getQueryString({
    ...params,
  });

  const { data } = await axios.get(`${ENDPOINT_API}/workers?${urlParameters}`, {
    headers,
  });
  return data;
};

export const getWorker = async (
  id: number,
  params?: Record<string, unknown>
): Promise<Worker> => {
  const data = await getWorkers({ id, ...params });
  return data[0];
};

export const getWorkerByEmail = async (
  email: string,
  params?: Record<string, unknown>
): Promise<Worker> => {
  const { data } = await axios.get(`${ENDPOINT_API}/workers`, {
    headers,
    params: { email, ...params },
  });
  return data[0];
};

export const addWorker = async (formData: Worker): Promise<Worker> => {
  const { data } = await axios.post(`${ENDPOINT_API}/workers`, formData, {
    headers: { ...headers, 'Content-Type': 'application/json' },
  });
  return data;
};

export const updateWorker = async (formData: Worker): Promise<Worker> => {
  await axios.patch(`${ENDPOINT_API}/workers`, formData, {
    headers: { ...headers, 'Content-Type': 'application/json' },
  });
  return formData;
};
