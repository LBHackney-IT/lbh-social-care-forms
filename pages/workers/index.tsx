import Seo from 'components/Layout/Seo/Seo';
import WorkerSearch from 'components/WorkerView/WorkerSearch';

const SearchWorkerPage = (): React.ReactElement => (
  <>
    <div>
      <Seo title="Search Worker" />
      <h1 className="govuk-fieldset__legend--l gov-weight-lighter govuk-expand-title">
        Manage Workers
      </h1>
      <p className="govuk-body">
        Before creating or updating a worker in this system, you will need to
        check if their email address already exists in the Social Care System.
      </p>
      <p className="govuk-body">
        Workers also need to be added to an appropriate Google Group which will
        give them the correct permissions to use this system. If this has not
        already been done, contact{' '}
        <a href="mailto:social-care.support@hackney.gov.uk">
          social-care.support@hackney.gov.uk
        </a>
        .
      </p>
    </div>
    <WorkerSearch />
  </>
);

export default SearchWorkerPage;
