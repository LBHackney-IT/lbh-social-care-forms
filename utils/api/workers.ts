import useSWR, { SWRResponse } from 'swr';
import { Worker, ErrorAPI, WorkerSearchResult } from 'types';
import { getQueryString } from 'utils/urls';
import axios from 'axios';

export const useWorkerById = (id: number): SWRResponse<Worker, ErrorAPI> =>
  useSWR(`/api/workers/${id}`);

export const useWorker = (
  params:
    | {
        email: string;
      }
    | undefined
): SWRResponse<Worker[], ErrorAPI> =>
  useSWR(params ? `/api/workers?${getQueryString(params)}` : null);

export const addWorker = async (
  formData: Record<string, unknown>
): Promise<Record<string, unknown>> => {
  const { data } = await axios.post(`/api/workers`, formData);
  return { ref: data?.id, data };
};

export const updateWorker = async (
  workerId: number,
  formData: Record<string, unknown>
): Promise<Record<string, unknown>> => {
  const { data } = await axios.patch(`/api/workers/${workerId}`, formData);
  return { ref: data?.id, data };
};

export const useWorkersSearch = (
  workerName: string
): SWRResponse<WorkerSearchResult[], ErrorAPI> =>
  useSWR(workerName ? `/api/workersearch?workerName=${workerName}` : null);
