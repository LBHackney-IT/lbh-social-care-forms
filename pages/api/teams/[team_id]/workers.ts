import { StatusCodes } from 'http-status-codes';

import { getWorkers } from 'lib/workers';
import { isAuthorised } from 'utils/auth';

import type { NextApiRequest, NextApiResponse, NextApiHandler } from 'next';
import { AxiosError } from 'axios';

const endpoint: NextApiHandler = async (
  req: NextApiRequest,
  res: NextApiResponse
) => {
  const user = isAuthorised(req);
  if (!user) {
    return res.status(StatusCodes.UNAUTHORIZED).end();
  }
  if (!user.isAuthorised) {
    return res.status(StatusCodes.FORBIDDEN).end();
  }
  switch (req.method) {
    case 'GET':
      try {
        const data = await getWorkers(req.query);
        res.status(StatusCodes.OK).json(data);
      } catch (error) {
        console.error(
          'Workers get error:',
          (error as AxiosError)?.response?.data
        );
        (error as AxiosError)?.response?.status === StatusCodes.NOT_FOUND
          ? res
              .status(StatusCodes.NOT_FOUND)
              .json({ message: 'Workers Not Found' })
          : res
              .status(StatusCodes.INTERNAL_SERVER_ERROR)
              .json({ message: 'Unable to get the Workers' });
      }
      break;

    default:
      res
        .status(StatusCodes.BAD_REQUEST)
        .json({ message: 'Invalid request method' });
  }
};

export default endpoint;
