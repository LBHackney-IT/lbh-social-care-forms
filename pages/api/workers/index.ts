import { StatusCodes } from 'http-status-codes';
import { isAuthorised } from 'utils/auth';

import { getWorkers, addWorker } from 'lib/workers';

import type { NextApiRequest, NextApiResponse, NextApiHandler } from 'next';
import { AxiosError } from 'axios';
import { apiHandler } from 'lib/apiHandler';

const endpoint: NextApiHandler = async (
  req: NextApiRequest,
  res: NextApiResponse
) => {
  const user = isAuthorised(req);
  if (!user) {
    return res.status(StatusCodes.UNAUTHORIZED).end();
  }
  if (!user.isAuthorised) {
    return res.status(StatusCodes.FORBIDDEN).end();
  }
  switch (req.method) {
    case 'GET':
      try {
        const data = await getWorkers(req.query);
        res.status(StatusCodes.OK).json(data);
      } catch (error) {
        console.error(
          'Workers gets an error:',
          (error as AxiosError)?.response?.data
        );
        (error as AxiosError)?.response?.status === StatusCodes.NOT_FOUND
          ? res
              .status(StatusCodes.NOT_FOUND)
              .json({ message: 'Worker Not Found' })
          : res
              .status(StatusCodes.INTERNAL_SERVER_ERROR)
              .json({ message: 'Unable to get the worker' });
      }
      break;

    case 'POST':
      try {
        const data = await addWorker(req.body);
        res.status(StatusCodes.OK).json(data);
      } catch (error) {
        console.error(
          'Workers gets an error:',
          (error as AxiosError)?.response?.data
        );
        res
          .status(StatusCodes.INTERNAL_SERVER_ERROR)
          .json({ message: 'Unable to add the worker' });
      }
      break;

    default:
      res
        .status(StatusCodes.BAD_REQUEST)
        .json({ message: 'Invalid request method' });
  }
};
export default apiHandler(endpoint);
