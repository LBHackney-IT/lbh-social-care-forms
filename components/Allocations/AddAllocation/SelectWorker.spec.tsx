import { render } from '@testing-library/react';

import SelectWorker from './SelectWorker';

import { mockedWorker } from 'factories/workers';

describe('SelectWorker component', () => {
  it('should render a worker', () => {
    const { asFragment } = render(
      <SelectWorker
        records={[mockedWorker]}
        callback={(value: any) => console.log(value)}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
