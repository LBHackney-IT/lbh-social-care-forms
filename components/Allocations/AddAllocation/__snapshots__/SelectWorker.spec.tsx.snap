// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectWorker component should render a worker 1`] = `
<DocumentFragment>
  <table
    class="govuk-table lbh-table"
    data-testid="residents-table"
  >
    <thead
      class="govuk-table__head"
    >
      <tr
        class="govuk-table__row"
      >
        <th
          class="govuk-table__header"
          scope="col"
        >
          Select
        </th>
        <th
          class="govuk-table__header"
          scope="col"
        >
          Name
        </th>
        <th
          class="govuk-table__header"
          scope="col"
        >
          Total allocations
        </th>
        <th
          class="govuk-table__header"
          scope="col"
        />
      </tr>
    </thead>
    <tbody
      class="govuk-table__body"
    >
      <tr
        class="govuk-table__row"
      >
        <td
          class="govuk-table__cell"
        >
          <div
            class="govuk-radios__item"
          >
            <input
              class="govuk-radios__input"
              id="1"
              name="workerId"
              type="radio"
              value="1"
            />
            <label
              class="govuk-label govuk-radios__label"
              for="1"
            >
              <span
                class="govuk-visually-hidden"
              >
                Foo Bar
              </span>
            </label>
          </div>
        </td>
        <td
          class="govuk-table__cell"
        >
          Foo Bar
        </td>
        <td
          class="govuk-table__cell"
        >
          <label
            for="1"
          >
            0
          </label>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
